core/src/connection/manager.rs

```diff
@@ -426,10 +426,11 @@ impl<'a, I, C> EstablishedEntry<'a, I, C> {
     /// > task _may not be notified_ if sending the event fails due to
     /// > the connection handler not being ready at this time.
     pub fn notify_handler(&mut self, event: I) -> Result<(), I> {
-        let cmd = task::Command::NotifyHandler(event);
+        let cmd = task::Command::NotifyHandler(event); // (*)
         self.task.get_mut().sender.try_send(cmd)
             .map_err(|e| match e.into_inner() {
-                task::Command::NotifyHandler(event) => event
+                task::Command::NotifyHandler(event) => event,
+                _ => unreachable!("by (*)")
```

Comment:
Apart from very simple cases, I think the message alone is rarely sufficient to troubleshoot a panic - you always need to take a look at the code immediately surrounding the line where the panic triggered. The wording of the message here just reflects the currently established practice of supplying a "proof" when `expect()`ing combined with my approach to short proofs. So in the "proof-style" this could also be `unreachable!("Because `cmd` is `Command::NotifyHandler`")`, but I personally don't find it useful to "describe" the surrounding code in the message in this way, rather than just referring to a marked line that states that same assumption. If we used a different practice, where `expect()`, `unreachable()` etc. messages were supposed to describe what went wrong rather than why it cannot happen, they may indeed tend to be a bit more helpful, e.g. "Expected Command::NotifyHandler, got ...". I still find it helpful to succinctly refer to the lines in the code that establish the assumptions for an `expect()` or `unreachable()` but that could of course also be achieved via regular comments, outside of the panic message. So if we don't follow the proof-as-panic-message here, this could be:

```rust
x => unreachable!("Expected `Command::NotifyHandker`, got {:?}", x); // by (*)
```

I wouldn't mind doing this, if you agree that it is an improvement. Or did you have something else in mind?

Reply:
Resolved by https://github.com/libp2p/rust-libp2p/pull/1619/commits/603cb2b29958170ba42375b6c5bd439274eb6a71.